src/modules/cmdpal/Microsoft.CmdPal.UI.ViewModels/FormViewModel.cs

```diff
@@ -0,0 +1,144 @@
+﻿// Copyright (c) Microsoft Corporation
+// The Microsoft Corporation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Text.Json;
+using AdaptiveCards.ObjectModel.WinUI3;
+using AdaptiveCards.Templating;
+using Microsoft.CmdPal.Extensions;
+using Microsoft.CmdPal.UI.ViewModels.Models;
+using Windows.Data.Json;
+using Windows.System;
+
+namespace Microsoft.CmdPal.UI.ViewModels;
+
+public partial class FormViewModel(IForm _form, IPageContext context) : ExtensionObjectViewModel(context)
+{
+    private readonly ExtensionObject<IForm> _formModel = new(_form);
+
+    // Remember - "observable" properties from the model (via PropChanged)
+    // cannot be marked [ObservableProperty]
+    public string TemplateJson { get; protected set; } = "{}";
+
+    public string StateJson { get; protected set; } = "{}";
+
+    public string DataJson { get; protected set; } = "{}";
+
+    public AdaptiveCardParseResult? Card { get; private set; }
+
+    public override void InitializeProperties()
+    {
+        var model = _formModel.Unsafe;
+        if (model == null)
+        {
+            return;
+        }
+
+        try
+        {
+            TemplateJson = model.TemplateJson();
+            StateJson = model.StateJson();
+            DataJson = model.DataJson();
+
+            AdaptiveCardTemplate template = new(TemplateJson);
+            var cardJson = template.Expand(DataJson);
+            Card = AdaptiveCard.FromJsonString(cardJson);
+        }
+        catch (Exception e)
+        {
+            // If we fail to parse the card JSON, then display _our own card_
+            // with the exception
+            AdaptiveCardTemplate template = new(ErrorCardJson);
+
+            // todo: we could probably stick Card.Errrors in there too
+            var dataJson = $$"""
+{
+    "error_message": {{JsonSerializer.Serialize(e.Message)}},
+    "error_stack": {{JsonSerializer.Serialize(e.StackTrace)}},
+    "inner_exception": {{JsonSerializer.Serialize(e.InnerException?.Message)}},
+    "template_json": {{JsonSerializer.Serialize(TemplateJson)}},
+    "data_json": {{JsonSerializer.Serialize(DataJson)}}
+}
+""";
+            var cardJson = template.Expand(dataJson);
+            Card = AdaptiveCard.FromJsonString(cardJson);
+        }
+
+        UpdateProperty(nameof(Card));
+    }
+
+    public void HandleSubmit(IAdaptiveActionElement action, JsonObject inputs)
+    {
+        if (action is AdaptiveOpenUrlAction openUrlAction)
+        {
+            _ = Launcher.LaunchUriAsync(openUrlAction.Url);
```

Comment:
Typically, this would be a View-layer operation, us being a Windows-TFM lets us 'cheat'. Also, I know sometimes there can be problems if these aren't called on the UI thread.

We could have a generic "OpenUrlMessage" that the main shell page knows how to handle and just does this there, then it can be used by anyone anywhere.